### PR TITLE
[SC-11574] Add exception option to businessDays functions

### PR DESCRIPTION
**Pull request by ksk5280**
TODO: 
* implement exceptions option for `addBusinessDays` and `subBusinessDays`.  
* Write tests, add more edge cases
* Double check logic for when date range starts/ends on exceptions.

### CODE REVIEW

File: src/differenceInBusinessDays/test.ts [hunk 1]
```diff
@@ -65,37 +65,241 @@ describe('differenceInBusinessDays', function () {
...
+            '01/15/22': false,
+            '01/16/22': false,
+          },
+        }
+      )
+
+      assert(result === 4)
+    })
+
+    it('can handle exceptions that fall on both of the argument dates', function () {
```
Comment (ksk5280) [inline src/differenceInBusinessDays/test.ts:222]: This test fails. I put the result as what I _think_ it should be, so I think if both days are exceptions then one extra day should be added, but I do not have that implemented in the logic.

File: src/differenceInBusinessDays/test.ts [hunk 2]
```diff
@@ -65,37 +65,241 @@ describe('differenceInBusinessDays', function () {
...
+          exceptions: {
+            '01/08/22': true,
+            '01/09/22': true,
+            '01/16/22': true,
+          },
+        }
+      )
+
+      // TODO: check what the expected result should be
+      assert(result === 7)
```
Comment (khalidwilliams) [inline src/differenceInBusinessDays/test.ts:239]: @skunka7 Is this the expected result?

File: src/addBusinessDays/index.ts [hunk 3]
```diff
@@ -54,25 +70,69 @@ export default function addBusinessDays(
...
+        exceptions[exceptionString] === true &&
+        !businessDays.includes(exceptionDate.getDay())
+      ) {
+        return true
+      } else if (
+        exceptions[exceptionString] === false &&
+        businessDays.includes(exceptionDate.getDay())
+      ) {
+        return true
+      }
```
Comment (chr1sjf0x) [inline src/addBusinessDays/index.ts]: thought (non-blocking): Looks like this logic can be simplified?
```suggestion
      if (
        exceptions[exceptionString] === !businessDays.includes(exceptionDate.getDay())
      ) {
        return true
      } 
```

Also generally, it feels like a few strategically placed comments explaining how this changes here work could be helpful.

File: src/addBusinessDays/test.ts [hunk 4]
```diff
@@ -132,12 +177,19 @@ describe('addBusinessDays', function () {
     )
   })
 
-  it('if custom businessDays are Monday and Wednesday, it returns the Thursday when 2 days is added on the Sunday', function () {
+  it('if custom businessDays are Monday and Wednesday, it returns the Wednesday when 2 days is added on the Sunday', function () {
```
Comment (chr1sjf0x) [inline src/addBusinessDays/test.ts:182]: question: Was this a typo? Should it also be updated on line 185 below?

File: src/addBusinessDays/test.ts [hunk 5]
```diff
@@ -57,6 +57,51 @@ describe('addBusinessDays', function () {
     assert.deepStrictEqual(result, new Date(2022, 0 /* Jan */, 17))
   })
 
+  it('can add exceptions to include days as businessDays that otherwise would not be included', function () {
```
Comment (chr1sjf0x) [inline src/addBusinessDays/test.ts]: thought: maybe a `describe` block would be nice?

File: src/differenceInBusinessDays/index.ts [hunk 6]
```diff
@@ -85,14 +87,59 @@ export default function differenceInBusinessDays(
...
-  return result === 0 ? 0 : result
+  // handle exceptions
+  let exceptionCount = 0
+  if (options.exceptions) {
+    const exceptionDates = Object.keys(options.exceptions)
+    exceptionDates.forEach((e) => {
+      const date = new Date(e)
+      if (!isValid(date)) return
+      // if date is within the left and right dates
+      if (isBefore(date, dateLeft) && isAfter(date, dateRight)) {
```
Comment (chr1sjf0x) [inline src/differenceInBusinessDays/index.ts:107]: thought: I wonder if we could extract/share this filtering logic of exceptions?

**Review by chr1sjf0x — APPROVED** [review]
Looking good overall.

Seems like there are a few opportunities for cleanup/clarifying comments.

File: src/addBusinessDays/test.ts (hunk 4 above)
Comment (khalidwilliams) [reply src/addBusinessDays/test.ts:182]: I assumed it was a typo, I couldn't figure out why the correct answer would be a non-working day. And yes, I missed updating it below.

File: src/addBusinessDays/index.ts (hunk 3 above)
Comment (khalidwilliams) [reply src/addBusinessDays/index.ts]: Yeah that's a lot more clean, I'll update!

File: src/addBusinessDays/test.ts (hunk 4 above)
Comment (khalidwilliams) [reply src/addBusinessDays/test.ts:182]: Oh and the 15th is indeed a Wednesday.

File: src/differenceInBusinessDays/index.ts (hunk 6 above)
Comment (khalidwilliams) [reply src/differenceInBusinessDays/index.ts:107]: The differenceInBusinessDays function has some extra edge cases that make the logic difficult to share. For now I think I'll go with these functions as is and try using them in GQL service.

File: src/addBusinessDays/index.ts [hunk 7]
```diff
@@ -27,19 +31,31 @@ export default function addBusinessDays(
...
   }
 ): Date {
   requiredArgs(2, arguments)
   const options = dirtyOptions || {}
+  const exceptions = options.exceptions || {}
   const businessDays =
     options.businessDays == null
       ? [1, 2, 3, 4, 5]
-      : options.businessDays.map(toInteger)
+      : options.businessDays.filter((number) => number < 7).map(toInteger)
```
Comment (blambillotte) [inline src/addBusinessDays/index.ts:43]: @chr1sjf0x @khalidwilliams  Looks like I'm too late on this comment, but I'm curious about the use of these filters on `businessDays`  - am I correct in assuming this is to protect against an invalid `day` index? If so I would think we'd want to deal with this by throwing an error explicitly with a helpful message rather than silently filtering them away?

Comment (khalidwilliams) [reply src/addBusinessDays/index.ts:43]: @blambillotte yep, that's the intention behind those lines. I'm not sure if the filtering here is following a convention from the other parts of the library (ie: if 1 and 8 are always seen as a a Monday), as that line came from Kimi. But it's a good question and a quick fix if not. Worth circling back to on Monday once everyone's online?

Comment (blambillotte) [reply src/addBusinessDays/index.ts:43]: Got it, yeah looks like the lib uses the convention of throwing a `RangeError` in the event of invalid input - some examples to check out: https://github.com/date-fns/date-fns/search?q=RangeError
